### PR TITLE
[SYSTEMML-2077] New eval builtin function

### PR DESCRIPTION
**Pull request by EdgarLGB**
Hi,
I made an another PR for the new eval built-in function. I made a code refactoring and fixed the code format problem. And I put all the commits into the master branch. Thanks for review.

Regards,
Guobao

### CODE REVIEW

**Comment by akchinSTC** [comment]
Can an Admin verify this patch?

File: src/main/java/org/apache/sysml/hops/NaryOp.java [hunk 1]
```diff
@@ -159,7 +159,8 @@ else if ( areDimsBelowThreshold() )
 		setRequiresRecompileIfNecessary();
 		
 		//ensure cp exec type for single-node operations
-		if ( _op == OpOpN.PRINTF )
+		//TODO eval function only supports CP execution type
```
Comment (mboehm7) [inline src/main/java/org/apache/sysml/hops/NaryOp.java]: please remove the TODO as eval will be be similar to all other function calls always be executed in CP.

File: src/main/java/org/apache/sysml/hops/ipa/InterProceduralAnalysis.java [hunk 2]
```diff
@@ -89,7 +89,7 @@
 	protected static final boolean INTRA_PROCEDURAL_ANALYSIS      = true; //propagate statistics across statement blocks (main/functions)	
 	protected static final boolean PROPAGATE_KNOWN_UDF_STATISTICS = true; //propagate statistics for known external functions 
 	protected static final boolean ALLOW_MULTIPLE_FUNCTION_CALLS  = true; //propagate consistent statistics from multiple calls 
-	protected static final boolean REMOVE_UNUSED_FUNCTIONS        = true; //remove unused functions (inlined or never called)
+	protected static final boolean REMOVE_UNUSED_FUNCTIONS        = false; //remove unused functions (inlined or never called)
 	protected static final boolean FLAG_FUNCTION_RECOMPILE_ONCE   = true; //flag functions which require recompilation inside a loop for full function recompile
```
Comment (mboehm7) [inline src/main/java/org/apache/sysml/hops/ipa/InterProceduralAnalysis.java]: Once we merge this in, we would need to make this conditional on the presence of eval functions. I'm happy to take this over, but if you want to give it a shot, you can try extending the existing function call graph to capture this information and create the IPA rewrite set accordingly.

File: src/main/java/org/apache/sysml/parser/BuiltinFunctionExpression.java [hunk 3]
```diff
@@ -379,6 +380,14 @@ public void validateExpression(HashMap<String, DataIdentifier> ids, HashMap<Stri
 		this.setOutput(output);
 		
 		switch (this.getOpCode()) {
+		case EVAL:
+			if (_args.length == 0) {
+				raiseValidateError("Function eval should provide at least one arguments.", false);
```
Comment (mboehm7) [inline src/main/java/org/apache/sysml/parser/BuiltinFunctionExpression.java]: Typo: "one arguments" -> "one argument, i.e., the function name".

File: src/main/java/org/apache/sysml/parser/BuiltinFunctionExpression.java [hunk 4]
```diff
@@ -1792,6 +1801,12 @@ else if ( functionName.equals("bitwShiftR") )
 			bifop = Expression.BuiltinFunctionOp.BITWSHIFTR;
 		else if ( functionName.equals("ifelse") )
 			bifop = Expression.BuiltinFunctionOp.IFELSE;
+		else if (functionName.equals("eval")) {
+			bifop = Expression.BuiltinFunctionOp.EVAL;
+			//Need to cast the first param "name of the referenced func" into a string identifier
+			DataIdentifier func = (DataIdentifier) paramExprsPassed.get(0).getExpr();
+			paramExprsPassed.get(0).setExpr(new StringIdentifier(func.toString(), func));
```
Comment (mboehm7) [inline src/main/java/org/apache/sysml/parser/BuiltinFunctionExpression.java]: Please try to avoid this explicit modification. This should not be necessary and would probably crash if the argument is an expression such as "foo" + X[i,1]. Maybe you can add a test for such expressions to your unit test.

File: src/main/java/org/apache/sysml/parser/DMLTranslator.java [hunk 5]
```diff
@@ -1272,6 +1272,10 @@ public void constructHops(StatementBlock sb)
 					if (!(target instanceof IndexedIdentifier)) {
 						//process right hand side and accumulation
 						Hop ae = processExpression(source, target, ids);
+						// if it is eval built-in func, add hop to output
+						if (ae instanceof NaryOp && ((NaryOp) ae).getOp().equals(OpOpN.EVAL)) {
+							output.add(ae);
+						}
```
Comment (mboehm7) [inline src/main/java/org/apache/sysml/parser/DMLTranslator.java]: Why was this necessary and not sufficiently handled by the default case for setting the output properties?

File: src/main/java/org/apache/sysml/parser/dml/DmlSyntacticValidator.java [hunk 6]
```diff
@@ -519,7 +532,8 @@ public void exitBuiltinFunctionExpression(BuiltinFunctionExpressionContext ctx)
 		String functionName = names[1];
 
 		ArrayList<ParameterExpression> paramExpression = getParameterExpressionList(ctx.paramExprs);
-
+		//convert the namespace for the function identifier arguments
+		convertNamespace(paramExpression);
```
Comment (mboehm7) [inline src/main/java/org/apache/sysml/parser/dml/DmlSyntacticValidator.java]: Any modification to the syntactic parser for DML would also need to go into the syntactic parser for PyDML. Hence, if possible, it would be good to leave the syntactic parsers untouched and move any preprocessing into the semantic parsing, i.e., parse tree validation.

File: src/main/java/org/apache/sysml/parser/dml/DmlSyntacticValidator.java [hunk 7]
```diff
@@ -507,6 +507,19 @@ private void castAsScalarDeprecationCheck(String functionName, ParserRuleContext
 		}
 	}
 
+	private void convertNamespace(ArrayList<ParameterExpression> params) {
+		for (ParameterExpression p : params) {
```
Comment (mboehm7) [inline src/main/java/org/apache/sysml/parser/dml/DmlSyntacticValidator.java]: Doing that for all parameters of all builtin functions seems dangerous as it has the potential to break user-provided string parameters as well. Please restrict these modifications to the first parameter of eval calls.

File: src/main/java/org/apache/sysml/runtime/instructions/cp/EvalBuiltinNaryCPInstruction.java [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+	}
+
+	@Override
+	public void processInstruction(ExecutionContext ec) throws DMLRuntimeException {
+		//1. get the namespace and func
+		CPOperand func = inputs[0];
+		String funcName = func.getName();
+		String namespace = null;
+		if (func.getName().contains("::")) {
+			String[] split = func.getName().split("::");
```
Comment (mboehm7) [inline src/main/java/org/apache/sysml/runtime/instructions/cp/EvalBuiltinNaryCPInstruction.java]: Please replace with `DMLProgram.splitFunctionKey` to have a single place where these symbols are defined.

File: src/main/java/org/apache/sysml/runtime/instructions/cp/EvalBuiltinNaryCPInstruction.java [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+				Matcher matcher = pattern.matcher(func.getName());
+				if (matcher.matches()) {
+					funcName = matcher.group(0);
+				}
+			} else {
+				funcName = funcSignature;
+			}
+		}
+		// bound the inputs to avoiding being deleted after the function call
+		CPOperand[] boundInputs = Arrays.asList(inputs).subList(1, inputs.length).toArray(new CPOperand[]{});
```
Comment (mboehm7) [inline src/main/java/org/apache/sysml/runtime/instructions/cp/EvalBuiltinNaryCPInstruction.java]: Use `Arrays.copyOfRange` instead.

File: src/main/java/org/apache/sysml/runtime/instructions/cp/EvalBuiltinNaryCPInstruction.java [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+		ArrayList<String> boundOutputNames = new ArrayList<>();
+		boundOutputNames.add(output.getName());
+		ArrayList<String> boundInputNames = new ArrayList<>();
+		for (CPOperand input : boundInputs) {
+			boundInputNames.add(input.getName());
+		}
+
+		//2. copy the created output matrix
+		MatrixObject outputMO = new MatrixObject(ec.getMatrixObject(output.getName()));
```
Comment (mboehm7) [inline src/main/java/org/apache/sysml/runtime/instructions/cp/EvalBuiltinNaryCPInstruction.java]: There is no need to create a copy or even obtain this output object.

File: src/main/java/org/apache/sysml/runtime/instructions/cp/EvalBuiltinNaryCPInstruction.java [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+		//4. convert the result to matrix
+		Data newOutput = ec.getVariable(output);
+		if (newOutput instanceof MatrixObject) {
+			return;
+		}
+		MatrixBlock mb = null;
+		if (newOutput instanceof ScalarObject) {
+			//convert scalar to matrix
+			mb = new MatrixBlock(1, 1, false);
+			mb.setValue(0, 0, ((ScalarObject) newOutput).getDoubleValue());
```
Comment (mboehm7) [inline src/main/java/org/apache/sysml/runtime/instructions/cp/EvalBuiltinNaryCPInstruction.java]: use `new MatrixBlock(((ScalarObject) newOutput).getDoubleValue())` instead.

File: src/main/java/org/apache/sysml/runtime/instructions/cp/EvalBuiltinNaryCPInstruction.java [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+		} else if (newOutput instanceof FrameObject) {
+			//convert frame to matrix
+			mb = DataConverter.convertToMatrixBlock(((FrameObject) newOutput).acquireRead());
+			ec.cleanupCacheableData((FrameObject) newOutput);
+		}
+		outputMO.acquireModify(mb);
+		outputMO.release();
+		ec.setVariable(output.getName(), outputMO);
```
Comment (mboehm7) [inline src/main/java/org/apache/sysml/runtime/instructions/cp/EvalBuiltinNaryCPInstruction.java]: Use `ec.setMatrixOutput(output.getName(), mb);` instead

File: src/main/java/org/apache/sysml/parser/BuiltinFunctionExpression.java (hunk 4 above)
Comment (EdgarLGB) [reply src/main/java/org/apache/sysml/parser/BuiltinFunctionExpression.java]: I have a question about the first argument of eval func. Should it be better to always have a string of func name as the first argument, i.e, eval("f1", X, y)? Because if it accepts any expression as the first argument, i.e, eval("f1" + X[i, 1], X, y), it will be complicated to deal with this expression of any type. And in fact, I think will can use it in an another way, i.e, eval("f1", X, y) + X[i, 1].

File: src/main/java/org/apache/sysml/runtime/instructions/cp/EvalBuiltinNaryCPInstruction.java (hunk 10 above)
Comment (EdgarLGB) [reply src/main/java/org/apache/sysml/runtime/instructions/cp/EvalBuiltinNaryCPInstruction.java]: I think that the eval func always returns a matrix. Then I set the ouput type with Matrix. So I copy the created output matrix object to avoid overriding it when the func call returns a scalar object.

**Comment by EdgarLGB** [comment]
Hi,
I have made the refactoring. And I'd like to consider the alias of the imported source as a macro. Hence, when meeting the namespace alias in symantic validation of eval func, I can replace the alias with the real complete path of the namespace.

File: src/main/java/org/apache/sysml/runtime/instructions/cp/EvalBuiltinNaryCPInstruction.java (hunk 10 above)
Comment (mboehm7) [reply src/main/java/org/apache/sysml/runtime/instructions/cp/EvalBuiltinNaryCPInstruction.java]: Yes, you right, I forgot about the scalar case. Since the copy constructor creates a different ID maybe you alternatively, get the object and disable its cleanup with `mo.enableCleanup(false)`, which preserves the contents but might drop it from the symbol table.

File: src/main/java/org/apache/sysml/parser/BuiltinFunctionExpression.java (hunk 4 above)
Comment (mboehm7) [reply src/main/java/org/apache/sysml/parser/BuiltinFunctionExpression.java]: Let us separate two aspects: the type of the input, and the script level restrictions. The type should indeed always be a string which also guarantees that you will always get a string during runtime. 

However, we should not artificially restrict the possibility to have arbitrary expressions as input at script level because the common case will be either selections eval(X[i,]) or concatenations eval("foo"+i). These expressions will be handled automatically by the compiler, as done for any other builtin function - so you don't need to worry about it. 

Similarly, we will also allow eval in expression, but as mentioned before I will add a rewrite that splits the DAGs after eval in order to create a hook for dynamic recompilation because the sizes (dimensions and sparsity) of an unknown function will be unknown. Without dynamic recompilation this would lead to unnecessary distributed operations. In your tests you did not encounter this issue because toString is only supported as in-memory operation.

**Comment by mboehm7** [comment]
ok thanks @EdgarLGB - once the validation of input parameters is generalized for expressions I can merge this in. Note that it might take a day or two before it'll be pushed to master because I need to add the other features first (conditionally disable function removal, rewrite for DAG split after eval).

**Comment by EdgarLGB** [comment]
ok thanks @mboehm7

**Comment by mboehm7** [comment]
LGTM - thanks @EdgarLGB I'll merge this in together with the additional features. 

Just to answer your question: In practice, a lot of scripts depend on dynamic recompilation to adapt runtime plans for initial unknowns or changing sizes (dimensions and sparsity). We compile at the granularity of entire HOP DAGs. Since the output of eval (which calls an unknown function) will in most cases be unknown during initial compilation, all operations with data dependencies on the eval output would fallback to conservative distributed operations. Hence, we want - similar as we do it for data-dependent operators like table or removeEmpty (see `RewriteSplitDagDataDependentOperators`) - cut the DAG after this operator to create a hook for dynamic recompilation. At the same time, we don't want to cut the DAG into too many pieces because this limits rewrite potential - accordingly, the above rewrite tries to find a minimal cut in case there are multiple relevant operators in the same DAG.

**Comment by mboehm7** [comment]
Thanks again @EdgarLGB for your contribution - it worked like a charm. I further added the support for function name expressions, but removed the support for function calls outside the default namespace for two reasons. First, there were potential conflicts with constant variables, which could have caused incorrect results and runtime failures. Second, the call to namespace functions would still be limited to static names, which could simply be called directly. However, I took a note and will extend that once we cleaned up the entire meta data management of namespaces. Thanks.

**Comment by EdgarLGB** [comment]
cool @mboehm7  And if we could support the global constant variables to resolve the conflicts?

**Comment by mboehm7** [comment]
@EdgarLGB no global constants will not solve this issue. Doing namespace resolution by putting the mapping into the data structures used for constant propagation could overwrite constants with the same variable name as the namespace (not that variables can have the same name as namespaces or even builtin functions). 

Originally, we used the logical names as namespace, but it was later changed to the filenames to address issues with recursive namespace and function resolution. However, down the road (after our SystemML 1.1 release), we should simply maintain this additional meta data in the function directory. Once this is done, we could simply allow namespace functions in eval as well.
